arroyo-worker/src/connectors/kinesis/source/mod.rs

```diff
@@ -0,0 +1,508 @@
+                            || err.is_provisioned_throughput_exceeded_exception()
+                        {
+                            // TODO: make retry behavior configurable
+                            if retries == 5 {
+                                bail!("failed after {} retries", retries);
+                            }
+                            retries += 1;
+                            tokio::time::sleep(Duration::from_millis(200 * (1 << retries))).await;
+                        } else {
+                            return Err(anyhow!(error));
+                        }
+                    }
+                    _ => return Err(anyhow!(error)),
+                },
+            }
+        }
+    }
+
+    async fn handle_get_records(
+        &mut self,
+        shard_id: String,
+        get_records: GetRecordsOutput,
+        ctx: &mut Context<(), T>,
+    ) -> Result<Option<BoxedFuture<AsyncNamedResult<AsyncResult>>>, UserError> {
+        let last_sequence_number = get_records.records().and_then(|records| {
+            records
+                .last()
+                .map(|record| record.sequence_number().unwrap().to_owned())
+        });
+
+        let next_shard_iterator = self.process_records(get_records, ctx).await?;
+        let shard_state = self.shards.get_mut(&shard_id).unwrap();
+
+        if let Some(last_sequence_number) = last_sequence_number {
+            shard_state.offset = KinesisOffset::SequenceNumber(last_sequence_number.to_string());
+        }
+
+        match next_shard_iterator {
+            Some(shard_iterator_id) => Ok(Some(self.next_read_future(shard_id, shard_iterator_id))),
+            None => {
+                shard_state.closed = true;
+                Ok(None)
+            }
+        }
+    }
+    async fn handle_need_new_iterator(
+        &mut self,
+        shard_id: String,
+    ) -> Result<Option<BoxedFuture<AsyncNamedResult<AsyncResult>>>, UserError> {
+        let get_shard_iterator_call = self.get_shard_iterator_call();
+        let shard_state = self.shards.get_mut(&shard_id).unwrap();
+        Ok(Some(
+            shard_state.update_shard_iterator(get_shard_iterator_call),
+        ))
+    }
+
+    fn get_shard_iterator_call(&self) -> GetShardIterator {
+        self.kinesis_client.as_ref().unwrap().get_shard_iterator()
+    }
+
+    async fn run_int(&mut self, ctx: &mut Context<(), T>) -> Result<SourceFinishType, UserError> {
+        let config = load_from_env().await;
+        self.kinesis_client = Some(KinesisClient::new(&config));
+        let mut futures = FuturesUnordered::new();
+        let get_shard_iterator_call = self.get_shard_iterator_call();
+        for shard in self.shards.values_mut() {
+            if shard.closed {
+                continue;
+            }
+            let get_shard_iterator_call = shard.set_shard_iterator(get_shard_iterator_call.clone());
+            futures.push(shard.update_shard_iterator(get_shard_iterator_call));
```

Comment:
Yeah, this whole thing became a mess because I was trying to use a "convenience" method that returned the unwrapped &KinesisClient. This meant I was running into issues anytime I called get_mut() on self.shard, as that is a mutable borrow on self.shard so you can't also use the result of any other Self function. Inlining the shard unwrap fixes all of this, which I just did.

Reply:
Yeah, this whole thing became a mess because I was trying to use a "convenience" method that returned the unwrapped &KinesisClient. This meant I was running into issues anytime I called get_mut() on self.shard, as that is a mutable borrow on self.shard so you can't also use the result of any other Self function. Inlining the shard unwrap fixes all of this, which I just did.